uniform_isogeny_primes.py

```diff
@@ -0,0 +1,102 @@
+"""isogeny_primes.py
+
+    Return finite list of isogeny primes attached to a number field.
+
+    ====================================================================
+
+    This file is part of Isogeny Primes.
+
+    Copyright (C) 2022 Barinder S. Banwait and Maarten Derickx
+
+    Isogeny Primes is free software: you can redistribute it and/or
+    modify it under the terms of the GNU General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+    The authors can be reached at: barinder.s.banwait@gmail.com and
+    maarten@mderickx.nl.
+
+    ====================================================================
+
+"""
+
+# Imports
+
+import argparse
+import logging
+
+from sage.all import Integer, prime_divisors
+
+from sage_code.strong_uniform_bounds import unif_bd, type_one_unif_primes
+
+
+def do_uniform_quadratic():
+
+    EPSILONS = [(0,4), (0,6), (0,8), (4,4), (4,6), (4,12)]
+
+    for eps in EPSILONS:
+        bd = Integer(unif_bd(2,eps))
+        logging.info(f"Isogeny primes for {eps} are {bd.prime_divisors()}")
+
+    type_one_bound = type_one_unif_primes(int(2))
+
+    logging.info(f"Type 1 uniform bound is {type_one_bound}")
+
+
+def do_uniform(d, eps):
+    tr_not_6_bd = Integer(unif_bd(d, eps))
+    logging.info(f"Trace not 6 mod 12 bound is {tr_not_6_bd}")
+    # logging.info(f"Trace not 6 mod 12 primes are {tr_not_6_bd.prime_divisors()}")
+
+    type_one_bound = type_one_unif_primes(d)
+
+    logging.info(f"Type 1 uniform bound is {type_one_bound}")
+
+def get_eps_from_input(tuple_as_str):
+    return tuple([Integer(t) for t in tuple_as_str.split(",")])
+
+
+def cli_handler(args):  # pylint: disable=redefined-outer-name
+
+    loglevel = logging.DEBUG if args.verbose else logging.INFO
+    logging.basicConfig(
+        format="%(asctime)s %(levelname)s: %(message)s",
+        datefmt="%H:%M:%S",
+        level=loglevel,
+    )
+    logging.debug("Debugging level for log messages set.")
+
+    eps = get_eps_from_input(args.eps)
+
+    if args.d == int(2):
+        do_uniform_quadratic()
+    else:
+        do_uniform(args.d, eps)
+
+
+
+if __name__ == "__main__":
```

Comment:
With these changes we now have:
```
 sage uniform_isogeny_primes.py 2 --eps 0,4
17:09:01 INFO: Isogeny primes for (0, 4) are [2, 3, 5, 11, 17, 23, 29, 41, 53]
Traceback (most recent call last):
  File "/Users/maartenderickx/code/isogeny-primes2/uniform_isogeny_primes.py", line 102, in <module>
    cli_handler(args)
  File "/Users/maartenderickx/code/isogeny-primes2/uniform_isogeny_primes.py", line 80, in cli_handler
    do_uniform_quadratic()
  File "/Users/maartenderickx/code/isogeny-primes2/uniform_isogeny_primes.py", line 46, in do_uniform_quadratic
    bd = Integer(unif_bd(2,eps))
  File "/Users/maartenderickx/code/isogeny-primes2/sage_code/strong_uniform_bounds.py", line 168, in unif_bd
    return core_loop(d, eps, aux_primes)
  File "/Users/maartenderickx/code/isogeny-primes2/sage_code/strong_uniform_bounds.py", line 151, in core_loop
    assert B_star == B
AssertionError
```
I think this means we are now (as I expected) encountering integers that are 0. It is good news in a certain sense that it takes away my worries from an earlier comment. But it does mean some work.

Reply:
aah, so this assertion was quite heavy handed, actually we should only be asserting that B is nonzero in the case that trace of epsilon is not zero mod 6; let me add that in and see what happens ...

Reply:
indeed, fixing this to only assert when trace is not 0 mod 6 now works as before

Reply:
which in particular means that we don't have a nice small list of primes for (0,6), and instead only have the very large additive bound

Reply:
We should (at least in theory and under GRH) be able to bring down this large additive bound to something reasonable, just as we were are able to bring down the momose type 2 bound down uniformly in the case d is odd. Everything should work out really similar, the main question though is wether we can actually do this in practice or wether the large additive bound is to large in this case to even consider attempting to bring it down.

Reply:
as discussed in person, we're not going to do this right now ...